### DISCUSSION TITLE
Upgrading Axios from 0.22. to 1.7.2

### DISCUSSION THREAD

**Discussion by nkamp** [axios/axios#6433] [Q&A]
Hello,

I'm not quite new to Vue en certainly Axios. I'm stepping into a project. 
I could solver certain things but now I'm running into trouble and I don't know how to get further, also due a lack of knowledge of Axios\Vue.
  
I've question about upgrading of Axios. I have Axios upgraded from 0.22. to 1.7.2 and I get the following error:
> ERROR in src/api/xxxxxxxxxxx.ts:268:8
TS2345: Argument of type '{ url: string; method: string; }' is not assignable to parameter of type 'InternalAxiosRequestConfig<any>'.
  Property 'headers' is missing in type '{ url: string; method: string; }' but required in type 'InternalAxiosRequestConfig<any>'.

Here is the code which is calling the http (get):

```
export const profileVestiging = (
  branchNumber: string,
  onError: (apiError: ApiError) => void
): Promise<AxiosResponse<any> | void> =>
  http({
    url: `${ENDPOINT}/profile-vestiging?branch_number=${branchNumber}`,
    method: "get",
    headers: { "Gir-Error-Handling": LOCAL_ERROR_HANDLING },
  })
    .then((response) => response.data)
    .catch((err) => handleError(err, onError));
```

And here is the http method which is called:

```
import axios, { AxiosError, AxiosPromise, AxiosResponse } from "axios";

// eslint-disable-next-line
export const http = <T = any>(
  config: axiosInstance,
  onError: () => Promise<void> = () => Promise.resolve()
): AxiosPromise<T> =>
  axiosInstance(config).catch(async (error) => {
    await handleOptimisticLockingError(error, onError);
    throw error;
  });

const axiosConfig = {
  baseURL: baseUrl() !== undefined ? baseUrl() : "",
  timeout: REQUEST_TIMEOUT,
  withCredentials: true,
};

const axiosInstance = axios.create(
  process.env.VUE_APP_DISABLE_CSRF === "true"
    ? axiosConfig
    : { xsrfCookieName: "GIR_CSRF", xsrfHeaderName: "Gir-Csrf", ...axiosConfig }
);
```

I had also AxiosInstance in the import line, but it is removed by webstorm because it is not used. 
The axiosInstance is of the type AxiosInstane type, but that should be InternalAxiosInstance I think? And then the signature is changed or...?

Can someone give a hint or do suggestion how to solve this? Or point me into a direction that I can get further.

![vue Axios issue](https://github.com/axios/axios/assets/5973092/aa7fbe7f-d0a1-4ec1-a4d1-cbe8e7715973)

**Comment by jasonsaayman** [comment]
`InternalAxiosRequestConfig` is meant for Axios internals/interceptors, not for the public request config you pass into an Axios instance.

Your wrapper should accept `AxiosRequestConfig`, and your created instance should be typed as `AxiosInstance`:

```ts
import axios, {
  AxiosInstance,
  AxiosPromise,
  AxiosRequestConfig
} from 'axios';

const axiosInstance: AxiosInstance = axios.create(
  process.env.VUE_APP_DISABLE_CSRF === 'true'
    ? axiosConfig
    : { xsrfCookieName: 'GIR_CSRF', xsrfHeaderName: 'Gir-Csrf', ...axiosConfig }
);

export const http = <T = any>(
  config: AxiosRequestConfig,
  onError: () => Promise<void> = () => Promise.resolve()
): AxiosPromise<T> =>
  axiosInstance(config).catch(async error => {
    await handleOptimisticLockingError(error, onError);
    throw error;
  });
```

The reason you are seeing `headers` as required is that `InternalAxiosRequestConfig` represents the normalized config Axios uses internally after defaults are merged, where `headers` is always present. Call-site config objects like `{ url, method }` should use `AxiosRequestConfig` instead.
